lib/experimental/class-wp-rest-navigation-fallbacks-controller.php

```diff
@@ -0,0 +1,147 @@
+ */
+class WP_REST_Navigation_Fallbacks_Controller extends WP_REST_Controller {
+
+	/**
+	 * The Post Type for the Controller
+	 *
+	 * @var string
+	 */
+	private $post_type;
+
+	/**
+	 * Constructs the controller.
+	 *
+	 * @since 5.9.0
+	 */
+	public function __construct() {
+		$this->namespace = 'wp-block-editor/v1';
+		$this->rest_base = 'navigation-fallbacks';
+		$this->post_type = 'wp_navigation';
+	}
+
+	/**
+	 * Registers the controllers routes.
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+
+		// Lists a single nav item based on the given id or slug.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_fallbacks' ),
+					'permission_callback' => array( $this, 'get_fallbacks_permissions_check' ),
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::READABLE ),
+				),
+				'schema' => array( $this, 'get_fallbacks_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Retrieves the fallbacks' schema, conforming to JSON Schema.
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_fallbacks_schema() {
+		if ( $this->schema ) {
+			return $this->add_additional_fields_schema( $this->schema );
+		}
+
+		$this->schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'navigation-fallbacks',
+			'type'       => 'object',
+			'properties' => array(
+				'id' => array(
+					'description' => __( 'The unique identifier for the Navigation Menu.', 'default' ),
+					'type'        => 'integer',
+					'context'     => array( 'view', 'edit', 'embed' ),
+					'readonly'    => true,
+				),
+			),
+		);
+
+		return $this->add_additional_fields_schema( $this->schema );
+	}
+
+	/**
+	 * Checks if a given request has access to read fallbacks.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_fallbacks_permissions_check( $request ) {
+
+		$post_type = get_post_type_object( $this->post_type );
+
+		// Getting fallbacks also requires creating `wp_navigation` posts.
+		if ( ! current_user_can( $post_type->cap->create_posts ) ) {
+			return new WP_Error(
+				'rest_cannot_create',
+				__( 'Sorry, you are not allowed to create Navigation Menus as this user.', 'default' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		// Getting fallbacks requires reading `wp_navigation` posts.
+		return $this->check_has_read_only_access( $request, $post_type );
+	}
+
+	/**
+	 * Gets the most appropriate fallback Navigation Menu.
+	 *
+	 * @return int|null The ID of the fallback Navigation Menu, or null if none exists.
+	 */
+	public function get_fallbacks() {
+		$post = WP_Navigation_Fallbacks_Gutenberg::get_fallback_menu();
```

Comment:
I'll implement this.

Note to self: see https://developer.wordpress.org/rest-api/extending-the-rest-api/controller-classes/.

Reply:
You also need all this stuff. 

https://github.com/WordPress/wordpress-develop/blob/e5057af16e1c425ecc82cf043d38c71b517d9cb7/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1971-L1973

Reply:
> As noted on slack, I would recommend adding embed links to this request.

Could we do this in a follow up? I've added it to `Todos` in the PR description along with the other followups spinning out of this PR.

> You also need all this stuff.

Looking at this it seems the expectation is that the endpoint should always return an object. Indeed this is what is specified by the schema I wrote.

Initially I thought that as we were only going to be returning an ID it would be ok to return an `int`. However, I think it's necessary to return an object in order to allow for `add_additional_fields_to_object`

https://developer.wordpress.org/rest-api/extending-the-rest-api/modifying-responses/#adding-custom-fields-to-api-responses

Am I right?

Reply:
Also I'm not clear on whether to return an array or a `WP_Post` here? I assume an array is more suitable because we're only returning a single `id` property? Can you confirm?

Reply:
Adding links should be trival. Add this in `prepare_item_for_response`

```
if ( rest_is_field_included( '_links', $fields ) || rest_is_field_included( '_embedded', $fields ) ) {
			$links = $this->prepare_links( $post );
                         $response->add_links( $links );
}
```

```
protected function prepare_links( $post ) {
		return array(
			'self'       => array(
				'href' => rest_url( rest_get_route_for_post( $post->ID ) ),
                                 'embeddable' => true,
			),
                 );
```

Reply:
> Adding links should be trival

Yes that looks relatively simple. Thank you for providing it.

My point is that it's additional code to:

- write
- review

...which is not absolutely necessary for this PR to land. As someone whose never implemented this before it's highly likely I'd have made an error, which would have resulted in more review cycles for everyone involved.

I appreciate it may seem pedantic (it's not the intention) but I'm really just trying to keep things as simple and focused as possible. This PR is already too big for my liking. 

Is this a blocker?

Reply:
Why not just add it, the work is written? We are spending more time talking about not doing this than just doing it.

Reply:
To summarize for future reviewers, there are some open points in this exchange waiting for input from @spacedmonkey 

- what should the endoint return best: an int or a WP_Post?
- why do we need to add links, how does it enhance this endpoint whcih is merely getting one specific fallback `wp_navigation` post type?

Also `prepare_item_for_response` will be implemented in a future commit in this PR by @getdave

Reply:
@draganescu Thanks for summarising. A few responses just to add clarity:

> what should the endpoint return best: an int or a WP_Post?

I believe it should return a data structure (`array` seems to be commonly used in REST responses) with an `id` prop. A `WP_Post` may be invalid as the majority of properties would be missing. Moreover, including an entire `WP_Post` might necessitate providing a full schema to cater for the _entire_ `WP_Post` object (see below for more on that).

> why do we need to add links, how does it enhance this endpoint whcih is merely getting one specific fallback wp_navigation post type?

Originally the endpoint returned the entire `wp_navigation` object as a `WP_Post`. Apart from being extra data over the wire, the main problem I encountered was that I had to provide the schema for the _entire_ `WP_Post` for the endpoint. 

I reached out to @spacedmonkey and @TimothyBJacobs because that approach didn't sit well with me. [They advised](https://wordpress.slack.com/archives/C02RQC26G/p1681740107995419) that this was an ideal use case for [Linking and Embedding](https://developer.wordpress.org/rest-api/using-the-rest-api/linking-and-embedding/). This would allow me to return _only_ an `id` from the endpoint and instead use `links` to include a reference to the full post object (available at the standard Core Rest Post endpoint).

It's worth noting that the JS implementation in this PR does not require the full post object. Only an `id`. Therefore this PR (which is already complex) doesn't need this additional complication. I've been [discussing options for including links in the response](https://github.com/WordPress/gutenberg/issues/49916) with folks and we've come up with some potential solutions which might allow us to take advantage of embedded posts to avoid a 2nd network request in the block. 

> Also prepare_item_for_response will be implemented in a future commit in this PR by @getdave

It's already implemented. I just need to clarify that it is done correctly.